### PR TITLE
Fix holodeck material exploits

### LINKED ISSUE #70942: The Holodeck Winter Wonderland can produce infinite wood
Reporting client version: 514.1589

## Round ID:
[193450](https://scrubby.melonmesa.com/round/193450)

## Testmerges:
- [Fix halloweens races](https://www.github.com/tgstation/tgstation/pull/70874)
- [JSON Savefiles | Player Saves use JSON](https://www.github.com/tgstation/tgstation/pull/70492)

## Reproduction:

1. set holodeck to winter wonderland
2. cut trees with hatchet
3. set holodeck to something else and to winter wonderland again
4. free wood

### LINKED ISSUE #71019: Chess holodeck simulation can produce infinite metal
## Reproduction:
1. set holodeck to chess
2. deconstruct chesspieces with wrench
3. set holodeck to something else, then chess again
4. repeat
5. profit

### PR DESCRIPTION
**Pull request by mc-oofert**
## About The Pull Request

Makes hologram trees unable to be harvested or uprooted (no free wood)
Makes hologram chess pieces unable to be deconstructed with a wrench (no free metal)

## Why It's Good For The Game
Fixes #71019
Fixes #70942
## Changelog
:cl:
fix: You can no longer deconstruct the chess pieces or cut the trees in the holodeck for free mats
/:cl:

### CODE REVIEW

File: code/game/objects/structures/chess.dm [hunk 1]
```diff
@@ -8,6 +8,8 @@
 	max_integrity = 100
 
 /obj/structure/chess/wrench_act(mob/user, obj/item/tool)
+	if(src.flags_1 & HOLOGRAM_1)
```
Comment (Fikou) [inline code/game/objects/structures/chess.dm:11]: don't do src.var when unnecessary (almost always)

File: code/game/objects/structures/flora.dm [hunk 2]
```diff
@@ -91,7 +91,7 @@ GLOBAL_LIST_EMPTY(flora_uprooting_tools_typepaths)
 	if(user.combat_mode)
 		return ..()
 
-	if(can_uproot && is_type_in_typecache(used_item, uprooting_tools))
+	if(can_uproot && is_type_in_typecache(used_item, uprooting_tools) && !(src.flags_1 & HOLOGRAM_1))
```
Comment (Fikou) [inline code/game/objects/structures/flora.dm:94]: same

File: code/game/objects/structures/flora.dm [hunk 3]
```diff
@@ -183,6 +183,8 @@ GLOBAL_LIST_EMPTY(flora_uprooting_tools_typepaths)
  */
 /obj/structure/flora/proc/can_harvest(mob/user, obj/item/harvesting_item)
 
+	if(src.flags_1 & HOLOGRAM_1)
```
Comment (Fikou) [inline code/game/objects/structures/flora.dm:186]: same

**Comment by msgerbs** [comment]
Can we get some feedback when you try rather than just nothing?

**Comment by mc-oofert** [comment]
> Can we get some feedback when you try rather than just nothing?

it should come natural to people that you probably cant make wood out of a hologram

**Comment by Mothblocks** [comment]
you should probably add feedback anyway, even if its just a balloon alert saying "your wrench goes right through!" or something.

also if returning false has behavior you dont like, why do you return false for the chess piece?

**Comment by mc-oofert** [comment]
> also if returning false has behavior you dont like, why do you return false for the chess piece?

what i meant to say is that hitting it with a wrench is feedback by me

**Comment by msgerbs** [comment]
IMO hitting it instead isn't a good enough feedback. And yeah, I know, it should be obvious but there's a _lot_ to know in this game and honestly sometimes people just blank on something. It's better (and more flavorful) to give at least some form of feedback pretty much every time something doesn't work the way it normally would.

**Comment by Mothblocks** [comment]
yeah thats pretty bad. make it a balloon alert

**Comment by mc-oofert** [comment]
2022-11-08 07_01_11-Window
2022-11-08 07_01_36-Window

File: code/game/objects/structures/chess.dm [hunk 4]
```diff
@@ -9,7 +9,8 @@
 
 /obj/structure/chess/wrench_act(mob/user, obj/item/tool)
 	if(flags_1 & HOLOGRAM_1)
-		return FALSE
+		balloon_alert(user,"[tool] passes right through [src]!")
```
Comment (Mothblocks) [inline code/game/objects/structures/chess.dm:12]:
```suggestion
		balloon_alert(user, "it goes right through!")
```

we avoid direct names in balloonies

File: code/game/objects/structures/flora.dm [hunk 5]
```diff
@@ -90,8 +90,10 @@ GLOBAL_LIST_EMPTY(flora_uprooting_tools_typepaths)
 /obj/structure/flora/attackby(obj/item/used_item, mob/living/user, params)
 	if(user.combat_mode)
 		return ..()
-
-	if(can_uproot && is_type_in_typecache(used_item, uprooting_tools) && !(flags_1 & HOLOGRAM_1))
+	if(flags_1 & HOLOGRAM_1)
+		balloon_alert(user,"[used_item] passes right through [src]!")
```
Comment (Mothblocks) [inline code/game/objects/structures/flora.dm:94]:
```suggestion
		balloon_alert(user, "it goes right through!")
```

**Review by Mothblocks — CHANGES_REQUESTED** [review]

**Review by Mothblocks — APPROVED** [review]

**Comment by Mothblocks** [comment]
@mc-oofert Merge from upstream

**Comment by mc-oofert** [comment]
> @mc-oofert Merge from upstream

done

**Pull request merged**